source/extensions/filters/http/jwt_authn/verifier.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include "envoy/http/async_client.h"
+
+#include "extensions/filters/http/jwt_authn/authenticator.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+class Verifier;
+typedef std::unique_ptr<Verifier> VerifierPtr;
+
+typedef std::function<AuthenticatorPtr(const std::vector<std::string>& audiences)> AuthFactory;
+
+/**
+ * Supports verification of JWTs with configured requirments.
+ */
+class Verifier {
+public:
+  virtual ~Verifier() {}
+  /**
+   * Object used to signal the caller of completion.
+   */
+  class Callbacks {
+  public:
+    virtual ~Callbacks() {}
+
+    virtual void onComplete(const ::google::jwt_verify::Status& status) PURE;
+  };
+  // Verify all tokens on headers, and signal the caller with callback.
+  virtual void verify(Http::HeaderMap& headers, Callbacks& callback) PURE;
+  // close resouces in use.
+  virtual void close() PURE;
+
+  // Factory method for creating verifiers.
+  static VerifierPtr
+  create(const ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement& requirement,
+         const Protobuf::Map<ProtobufTypes::String,
+                             ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider>&
+             providers,
+         const AuthFactory& factory);
+};
+
+// base verifier for provider_name, provider_and_audiences, and allow_missing_or_failed.
+class AuthVerifier : public Verifier, public Authenticator::Callbacks {
```

Comment:
Only Verifier is an interface.  all others are its implementation.  so naming wise, how about
BaseVerifierImpl
ProviderNameVerifierImpl
xxxVeriferImpl

Reply:
Renamed implementations and moved them to code file.